Comply with 1.0.1 and 1.1.0.

anokfireball opened the pull request:
Gracefully andle the undefined case when meshesPath / particlesPath are
not present as paths, but are in the root path as attributes (1.0.1).
Also enfoce the new 1.1.0 rule that enforces a path when the attribute
exists.

anokfireball committed:
Comply with 1.0.1 and 1.1.0.

Gracefully andle the undefined case when meshesPath / particlesPath are
not present as paths, but are in the root path as attributes (1.0.1).
Also enfoce the new 1.1.0 rule that enforces a path when the attribute
exists.

anokfireball commented:
Fixes #48

ax3l commented on .travis/download_samples.sh:
```diff
@@ -7,3 +7,13 @@ tar -xf example-3d.tar.gz
+wget -nv https://github.com/yt-project/yt/files/1542670/no_particles.zip
+unzip no_particles.zip
+mv no_particles samples/issue-sample/
+rm -rf no_fields.zip no_particles.zip;
```
missing EOF empty newline :)

anokfireball committed:
Optional attributes from 1.1.0.

Support softwareDependencies and machine idiomatically.
Improve handling of meshesPath and particlesPath.

ax3l commented on src/Iteration.cpp:
```diff
@@ -148,15 +148,29 @@ Iteration::flush()
-    meshes.flush(o->meshesPath());
-    for( auto& m : meshes )
-        m.second.flush(m.first);
+    //TODO warn if openPMD >= 1.1.0, mp is set and no meshes
```
warning on a valid meshesPath without meshes is not necessary, it's just zero meshes.

the only thing that 1.1.0 enforces is to throw an error if meshesPath is set and the path is invalid (group does not exist)

the same for particles below

anokfireball replied:
Brain was not fully working at 7AM yet.
When writing the TODO, I was mislead into thinking the path gets created unconditionally in ```Series```, but it does 6 lines below.

ax3l commented on test/CoreTest.cpp:
```diff
@@ -86,35 +84,23 @@ BOOST_AUTO_TEST_CASE(output_default_test)
+    BOOST_TEST(o.iterationEncoding() == IE::groupBased);
+    BOOST_TEST(o.iterationFormat() == "/data/%T/");
+    BOOST_TEST(o.iterations.size() == 1);
+    BOOST_TEST(o.numAttributes() == 7); /* openPMD, openPMDextension, basePath, meshesPath, particlesPath, iterationEncoding, iterationFormat */
```
it's always allowed to have more attributes than openPMD defines, so this is a sub-ideal test. maybe make it `>=7`?

anokfireball replied:
I don't see how this is sub-ideal. ```numAttributes``` counts the number of currently set Attributes. As this test happens in a very conttrolled environment, we should check precisely whether the defaults reflect the standard in that environment.

ax3l replied:
oh wait, it's a write test not a read test. you are right!

ax3l reviewed (approved):
thx, all right!

I added some comments that can be addressed in a follow-up PR.
